# Задание №5 (Анализ и построение алгоритмов для исполнителей)

---

- Сложность: легкое
- Время выполнения: 4 минуты

---


## 1. Необходимый теоретический минимум

### 1.1. Системы счисления

> Система счисления (СС) — это способ записи чисел с помощью определённого набора цифр и правил, по которым определяется значение записанного числа.

Каждая система счисления характеризуется:

- Основанием (базой) — количеством различных цифр, которые в ней используются.
- Цифрами — символами, которые допустимы в записи числа.
- Позиционностью — значение каждой цифры зависит от её позиции (разряда) в числе.

| Название          | Основание | Цифры                 |
| ----------------- | --------- | --------------------- |
| Двоичная          | 2         | 0, 1                  |
| Восьмеричная      | 8         | 0–7                   |
| Десятичная        | 10        | 0–9                   |
| Шестнадцатеричная | 16        | 0–9 и A (10) – F (15) |

#### 1.1.1 Работа с системами счисления в Python

В языке Python удобно работать с числами в разных системах счисления. Вот основные способы:

---

**Перевод из десятичной системы**:

| Цель                    | Функция      | Пример     | Результат  |
| ----------------------- | ------------ | ---------- | ---------- |
| В **двоичную**          | `bin(число)` | `bin(10)`  | `'0b1010'` |
| В **восьмеричную**      | `oct(число)` | `oct(10)`  | `'0o12'`   |
| В **шестнадцатеричную** | `hex(число)` | `hex(255)` | `'0xff'`   |

*Примечание*: Префиксы `0b`, `0o`, `0x` — это обозначения систем:

- `0b` — бинарная (двоичная),
- `0o` — октальная (восьмеричная),
- `0x` — хекс (шестнадцатеричная).

---

#### Перевод **в десятичную** систему:

Используется встроенная функция `int()`:

```python
int('1010', 2)     # → 10 (из двоичной)
int('12', 8)       # → 10 (из восьмеричной)
int('ff', 16)      # → 255 (из шестнадцатеричной)
```

`int(строка, основание)` — переводит строку из заданной системы счисления в десятичную.

---

#### 1.1.2 Алгоритмы перевода из одной СС в другую СС

Переводы между системами счисления — важный навык для ЕГЭ по информатике. Ниже представлены понятные алгоритмы для ручного выполнения переводов.

---

**1. Из другой СС в десятичную**

**Алгоритм:**

1. Запиши число.
2. Каждую цифру умножь на основание системы в степени, соответствующей позиции цифры (счёт справа налево с нуля).
3. Сложи все полученные значения.

**Пример:**
Перевести $1011_2$ в десятичную:

$$
1 \cdot 2^3 + 0 \cdot 2^2 + 1 \cdot 2^1 + 1 \cdot 2^0 = 8 + 0 + 2 + 1 = 11
$$

---

**2. Из десятичной в другую СС**

**Алгоритм (делением с остатком):**

1. Делим число на основание новой системы счисления.
2. Запоминаем остаток.
3. Делим частное снова на основание — и так до тех пор, пока результат деления не станет 0.
4. Ответ — запись остатков **в обратном порядке**.

**Пример:**
Перевести $25_{10}$ в двоичную:

$$
25 \div 2 = 12 \text{ (остаток 1)}  
$$

$$
12 \div 2 = 6 \text{ (остаток 0)}  
$$

$$
6 \div 2 = 3 \text{ (остаток 0)}  
$$

$$
3 \div 2 = 1 \text{ (остаток 1)}  
$$

$$
1 \div 2 = 0 \text{ (остаток 1)}
$$

Ответ: $11001_2$

---

**3. Из одной СС в другую СС (не десятичную)**

1. Сначала переводи **в десятичную** (см. шаг 1).
2. Затем из десятичной — в нужную (см. шаг 2).

**Пример:**
Перевести $2F_{16}$ в двоичную:

- $2F_{16} = 2 \cdot 16 + 15 = 47_{10}$
- $47_{10} = 101111_2$

Ответ: $101111_2$


### 1.2 Бит четности

Последний справа бит у числа показывает его четность. Если стоит $1$ - нечетное число, а если $0$ - четное число.

In [7]:
for num in range(1, 10):
    print(f'Число {num} является {"  четным" if num % 2 == 0 else "нечетным"}, т.к. в конце двоичной записи ({bin(num)[2:]:4}) стоит -> {num % 2}')

Число 1 является нечетным, т.к. в конце двоичной записи (1   ) стоит -> 1
Число 2 является   четным, т.к. в конце двоичной записи (10  ) стоит -> 0
Число 3 является нечетным, т.к. в конце двоичной записи (11  ) стоит -> 1
Число 4 является   четным, т.к. в конце двоичной записи (100 ) стоит -> 0
Число 5 является нечетным, т.к. в конце двоичной записи (101 ) стоит -> 1
Число 6 является   четным, т.к. в конце двоичной записи (110 ) стоит -> 0
Число 7 является нечетным, т.к. в конце двоичной записи (111 ) стоит -> 1
Число 8 является   четным, т.к. в конце двоичной записи (1000) стоит -> 0
Число 9 является нечетным, т.к. в конце двоичной записи (1001) стоит -> 1


### 1.3 Удвоение числа в двоичной системе

Если к двоичному числу **дописать 0 справа**, оно **увеличится в 2 раза**.

Это похоже на десятичную систему:
добавление нуля справа к числу 23 → `230`, то есть умножение на 10.
А в двоичной системе добавление 0 → умножение на 2.

Примеры:

| Исходное | Добавили 0 | Результат | Десятичный смысл |
| -------- | ---------- | --------- | ---------------- |
| `101`    | `1010`     | 10        | 5 → 10           |
| `110`    | `1100`     | 12        | 6 → 12           |
| `11`     | `110`      | 6         | 3 → 6            |

**Запомни**: в Python то же самое можно сделать сдвигом:

In [8]:
n = 5
print(n << 1)  # результат: 10

10


### 1.4 Удаление последней цифры в двоичной записи

Чтобы **удалить последний бит справа**, нужно просто **разделить число на 2 нацело** (отбросить остаток).

Это эквивалентно **сдвигу вправо** в битовой арифметике.

Примеры:

| Двоичное | Удалили последнюю | Осталось | Десятичный смысл |
| -------- | ----------------- | -------- | ---------------- |
| `1101`   | → `110`           | 6        | 13 → 6           |
| `100`    | → `10`            | 2        | 4 → 2            |
| `1011`   | → `101`           | 5        | 11 → 5           |


In [9]:
n = 13
print(n // 2)   # результат: 6
# или
print(n >> 1)   # сдвиг вправо: тоже 6

6
6
